src/main/java/carpet/api/settings/SettingsManager.java

```diff
@@ -0,0 +1,796 @@
+            }
+        }
+    }
+
+    /**
+     * Calling this method is not supported.
+     */
+    public void inspectClientsideCommand(CommandSourceStack source, String string)
+    {
+        if (string.startsWith("/" + identifier + " "))
+        {
+            String[] res = string.split("\\s+", 3);
+            if (res.length == 3)
+            {
+                String setting = res[1];
+                String strOption = res[2];
+                if (rules.containsKey(setting) && rules.get(setting).canBeToggledClientSide())
+                {
+                    try {
+                        rules.get(setting).set(source, strOption);
+                    } catch (InvalidRuleValueException e) {
+                        e.notifySource(source);
+                    }
+                }
+            }
+        }
+    }
+    
+    private void switchScarpetRuleIfNeeded(CommandSourceStack source, CarpetRule<?> carpetRule) //TODO remove. This should be handled by the rule
+    {
+        if (carpetRule instanceof ParsedRule<?> rule && !rule.scarpetApp.isEmpty())
+        {
+            if (RuleHelper.getBooleanValue(rule) || (rule.type() == String.class && !rule.value().equals("false")))
+            {
+                CarpetServer.scriptServer.addScriptHost(source, rule.scarpetApp, s -> CommandHelper.canUseCommand(s, rule.value()), false, false, true, null);
+            } else {
+                CarpetServer.scriptServer.removeScriptHost(source, rule.scarpetApp, false, true);
+            }
+        }
+    }
+
+    private Path getFile()
+    {
+        return server.getWorldPath(LevelResource.ROOT).resolve(identifier + ".conf");
+    }
+    
+    private Collection<CarpetRule<?>> getRulesSorted()
+    {
+        return rules.values().stream().sorted(comparing(CarpetRule::name)).toList();
+    }
+
+    /**
+     * Disables all {@link CarpetRule}s with the {@link RuleCategory#COMMAND} category,
+     * called when the {@link SettingsManager} is {@link #locked}.
+     */
+    private void disableBooleanCommands()
+    {
+        for (CarpetRule<?> rule : rules.values())
+        {
+            if (!rule.categories().contains(RuleCategory.COMMAND))
+                continue;
+            try {
+                if (rule.suggestions().contains("false"))
+                    rule.set(server.createCommandSourceStack(), "false");
+                else
+                    CarpetSettings.LOG.warn("Couldn't disable command rule "+ rule.name() + ": it doesn't suggest false as a valid option");
+            } catch (InvalidRuleValueException e) {
+                throw new IllegalStateException(e); // contract of CarpetRule.suggestions()
+            }
+        }
+    }
+
+    private void writeSettingsToConf(ConfigFileData data)
+    {
+        if (locked)
+            return;
+        try (BufferedWriter fw = Files.newBufferedWriter(getFile()))
+        {
+            for (String comment : data.comments())
```

Comment:
That's why I think worrying about commands is not what we should worry at this point. If you have mixed in rules and comments they will be all smooshed together at this point. Its better IMO to be explicit that comments are preserved only when config is locked. If its locked its manually curated, so comments make sense. If they are automatically rendered, this means comments should not exist anyways, since its the game that edits them.

If we would want to do this properly saving the rules should be done via re-writing of the conf, skipping rules that are removed, updating rules that changed and appending new rules at the end if they are added. But that would definitely be a different PR

Reply:
I totally forgot I kept this in here when moving the class between packages, I'll revert it back.

Reply:
Removed saving comments for later, kept ignoring them so it can work for locked configs and default files.